modules/mqtt/testcontainers/mosquitto/__init__.py

```diff
@@ -0,0 +1,247 @@
+
+        interval = 1.0
+        timeout = 5
+        start = time.time()
+        while True:
+            duration = time.time() - start
+            if client.is_connected():
+                return
+            if duration > timeout:
+                raise TimeoutError(f"Failed to estabilish a connection after {timeout:.3f} " "seconds")
+            # wait till secondary thread manages to connect successfully:
+            time.sleep(interval)
+
+    def get_client(self, **kwargs) -> tuple[mqtt_client.Client, paho.mqtt.enums.MQTTErrorCode]:
+        """
+        Get a paho.mqtt client connected to this container.
+        Check the returned object is_connected() method before use
+
+        Args:
+            **kwargs: Keyword arguments passed to `paho.mqtt.client`.
+
+        Returns:
+            client: MQTT client to connect to the container.
+            error: an error code or MQTT_ERR_SUCCESS.
+        """
+        err = paho.mqtt.enums.MQTTErrorCode.MQTT_ERR_SUCCESS
+        if self.client is None:
+            self.client = mqtt_client.Client(
+                client_id=MosquittoContainer.TESTCONTAINER_CLIENT_ID,
+                callback_api_version=mqtt_client.CallbackAPIVersion.VERSION2,
+                userdata=self,
+                **kwargs,
+            )
+            self.client._connect_timeout = 1.0
+
+            # connect() is a blocking call:
+            err = self.client.connect(self.get_container_host_ip(), int(self.get_exposed_port(self.port)))
+            self.client.on_message = MosquittoContainer.on_message
+            self.client.loop_start()  # launch a thread to call loop() and dequeue the message
+
+        return self.client, err
+
+    def start(self) -> Self:
+        super().start()
+        self._connect()
+        return self
+
+    def stop(self, force=True, delete_volume=True) -> None:
+        self.client.disconnect()
+        self.client = None  # force recreation of the client object at next start()
+        self.watched_topics = {}  # clean all watched topics as well
+        super().stop(force, delete_volume)
+
+    class WatchedTopicInfo:
+        def __init__(self):
+            self.count = 0
+            self.timestamp_start_watch = time.time()
+            self.last_payload = ""
+
+        def on_message(self, msg: mqtt_client.MQTTMessage):
+            self.count += 1
+            # for simplicity: assume strings are used in integration tests and are UTF8-encoded:
+            self.last_payload = msg.payload.decode("UTF-8")
+
+        def get_count(self):
+            return self.count
+
+        def get_last_payload(self):
+            return self.last_payload
+
+        def get_rate(self):
+            duration = time.time() - self.timestamp_start_watch
+            if duration > 0:
+                return self.count / duration
+            return 0
+
+    def on_message(client: mqtt_client.Client, mosquitto_container: "MosquittoContainer", msg: mqtt_client.MQTTMessage):
```

Comment:
this is quite a tight coupling, probably too much for a testing utility

Reply:
I understand your point of view. OTOH the reason for using test-containers is to run integration tests, right?
To do integration testing for MQTT, I think it's a basic necessity to be able to inspect what the client has published, isn't it?
I mean: whoever is going to use this test-container will need to have some functionality like this one right?
Of course I'm open to change it to be more generic, if there's a way to do it 
(e.g. I can change the code so that there is no need to explicitly "watch" a topic -- MosquittoContainer might just be creating a WatchedTopicInfo instance for every possible topic where traffic is received)...
What's your thought on that?

Reply:
I mean I am using this as my reference - https://github.com/testcontainers/testcontainers-java/blob/main/modules/hivemq/src/main/java/org/testcontainers/hivemq/HiveMQContainer.java - will have to review and reconsider

Reply:
This sounds a bit conservative but I'm not sure I want to tightly couple folks' workflow with testcontainers - like you should still have very good logging in your application because hopefully, your application will make it to production, and then it will be up to cloudwatch/kibana/grafana to help you debug. if you were debugging this whole time with testcontainers utility methods, now you are going to be in a tight spot.